### ISSUE TITLE
Allow more awesome arrow function usage

### ISSUE THREAD

**Issue by novemberborn** [dojo/meta#45]
The style guide is rather sour on arrow functions:

> Arrow functions MUST include parentheses around parameters and curly braces around the function body (i.e. implicit returns aren't allowed).

There was quite an [uproar in the `standard` community](https://github.com/feross/standard/issues/414) when parentheses became required in v6. This was reverted in v7. Leaving aside the desired for minimalism amongst those who use `standard` (myself included, I don't need no semi-colons! 😉 ), writing lean arrow functions is pretty great for functional code and promises:

```js
myArr.map(str => str.toUpperCase());

myPromise.then(arr => arr[0]).then(firstResult => {
  // do stuff
  return result;
});
```

versus:

```js
myArr.map((str) => {
  return str.toUpperCase();
});

myPromise.then((arr) => {
  return arr[0];
}).then((firstResult) => {
  // do stuff
  return result;
});
```

Except that even that last example isn't allowed, because:

> Arrow functions SHOULD NOT be used simply as function shorthand; they should ONLY be used specifically for execution context preservation.

So then we get:

```js
myArr.map(function(str) {
  return str.toUpperCase();
});

myPromise.then(function(arr) {
  return arr[0];
}).then(function(firstResult) {
  // do stuff
  return result;
});
```

There's already a [bunch](https://github.com/dojo/core/blob/3e903cdae782de56903dac506b95b5e2bf2da1f4/src/has.ts#L180) [of](https://github.com/dojo/core/blob/3e903cdae782de56903dac506b95b5e2bf2da1f4/src/request.ts#L16) [violations](https://github.com/dojo/core/blob/3e903cdae782de56903dac506b95b5e2bf2da1f4/src/Symbol.ts#L176) in `dojo-core` for this rule.

Personally I think everything should be allowed, whether that's using plain `function`, or arrow functions, with or without parentheses. It mostly depends on the code and strict rules are [hard to enforce](https://github.com/dojo/meta/issues/43).

**Comment by kitsonk** [comment]
For the first point, my opinion is that we should still be explicit.  Just like ASI, sometimes they are necessary, so they should be there.  I just still think:

```ts
/* this */
foo((x) => x * 2);
/* is better than */
foo(x => x * 2);
/* for readability, especially because */
foo(() => 2);
/* and */
foo((item, idx) => [ item, idx ]);
/* require parens */
```

But personally I can be persuaded otherwise.

As far as the second one, I know that is our standard, and I personally have been breaking it left and right and not really feeling at all guilty about it.  I think general use of lambda functions, even when canonical `this` is not required, are better used.  I would be all for rethinking that one.

For example, I think [this](https://github.com/dojo/compose/blob/7bfd731022f391f69622a7a051861922490f7c52/src/mixins/createStateful.ts#L184) is far more readable, though there is no use of canonical `this`:

```ts
observedState = {
	id,
	observable,
	subscription: observable
		.observe(id)
		.subscribe(
			(item) => setStatefulState(stateful, item), /* next handler */
			(err) => {
				/* TODO: Should we emit an error, instead of throwing? */
				throw err;
			}, /* error handler */
			() => unobserve(stateful)), /* completed handler */
	handle: {
		destroy() {
			const observedState = observedStateMap.get(stateful);
			if (observedState) {
				observedState.subscription.unsubscribe();
				observedStateMap.delete(stateful);
			}
		}
	}
}
```

**Comment by novemberborn** [comment]
> For the first point, my opinion is that we should still be explicit. Just like ASI, sometimes they are necessary, so they should be there.

Heh, one of the arguments for not enforcing them in standard was that semicolons are not enforced, unless necessary 😉 

Though you'd be OK with implicit return?

**Comment by kfranqueiro** [comment]
To channel rands, when you say "more awesome", I hear "less readable" ;)

One thing to bear in mind is that if you need to include typings with your arguments, [you need the parens anyway even for single-argument](https://github.com/Microsoft/TypeScript/issues/3323).

I'd also argue that the parentheses are more apt to signal to someone reading code they are unfamiliar with that "there is a function here".  I'd argue the same for including curly braces.

RE no braces / implicit return, here are some things to consider:

* Implicit return is really easy for untrained eyes to overlook, and I expect it to easily lead to subtle bugs.
* Implicit return is only applicable in the no-braces case, so once you need multiple statements, you need a `return` anyway.
* We already require curly braces around other single-statement blocks like if/else, so arguably requiring curly braces for arrow functions is consistent and provides a similar safety net (namely when editing code and suddenly going from single-statement to multi-statement...  rather easy to forget to add a return in that case, or end up asking yourself "hm, do consumers of this function *expect* me to return this or was I just implicitly returning because it's convenient and YOLO")
* Just as Kit is arguing that always including parens regardless of number of arguments is consistent, I could argue the same for always including curly braces regardless of number of statements

**Comment by kitsonk** [comment]
> * Just as Kit is arguing that always including parens regardless of number of arguments is consistent, I could argue the same for always including curly braces regardless of number of statements

The only counter argument is that it isn't "just braces":

```typescript
foo((x) => x * 2);
/* is different from */
foo((x) => { x * 2; });
/* to get the same */
foo((x) => { return x * 2; });
/* which we shouldn't do, but instead */
foo((x) => {
    return x * 2;
});
```

My feeling was because there isn't as easy to shoot yourself in the foot with single line lambdas and the intent is clear, my opinion was willing to not point that out, thought someone would. 😄 For some reason, though I thought I was going to feel dirty using them, I think it actually makes things more readable.  Especially when doing chaining, lambdas with implicit returns can actually make hard to follow code, much easier to follow.  But again, one persons opinion and I respect the counter arguments being made.

**Comment by msssk** [comment]
Bit of an aside, but I would not look to the standard community for guidance. I am not impressed with their very poor choices.

* single-line `if` statements
* no semi-colons
* spaces for indentation

**Comment by mwistrand** [comment]
While it is certainly true that some people give precious little thought to how they wield the newest tools placed into their hands, at the end of the day I do not think arrow functions have a detrimental effect on the code's readability. As @kitsonk has already pointed out, they can even aid readability when chaining:

```javascript
function fetchAllTheThings(urls, options) {
  const requests = urls.map(url => request.get(url, options));
  return Promise.all(requests)
    .then(results => parse(results))
    .then(parsed => {
      return parsed.map(data => createStore(data));
    });
}
```

In fact, a lot of developers already do something similar with functions:

```javascript
function fetchAllTheThings(urls, options) {
  var requests = urls.map(function (url) {
    return request.get(url, options);
  });
  
  return Promise.all(requests)
    .then(parseResults)
    .then(createStores);
}
```

Only instead of asking fellow developers to take the time to understand the functions being used, they are being asked to take the time to understand the language syntax being used (although I will admit that that comparison is not entirely apples-to-oranges). Of course, I am not saying we should get crazy with this:

```javascript
// this is awful
somePromise.then(results => results.map(result => result.id));
// and instead should be
somePromise.then(results => {
  return results.map(result => result.id);
});
```

As for requiring parens around the parameters, my personal preference is to keep them for the reasons already stated above, but ultimately there are other hills I am more keen to die on.

> Arrow functions SHOULD NOT be used simply as function shorthand; they should ONLY be used specifically for execution context preservation.

Personally, I wonder if the main argument for continuing to prefer functions whose default context is the global object (as opposed to something that makes more sense, like the current context) is that it is what everyone is used to. With this style of writing functions, you have to know how the function is being called in order to know what `this` represents. In opposition to this particular guideline, the benefit of favoring arrow functions in such scenarios is that seeing the `function` keyword can actually make it clearer that the context is being overwritten by the executor. As such, I am perfectly fine with allowing arrow functions even when context preservation is not needed.

**Comment by kitsonk** [comment]
We have generally had good conversation about this and have not come up with strict and hard rules.  We have agreed to allow arrow functions in callbacks, where it is essentially a "literal function" and avoided them in other situations.  We will re-open and revisit if we feel there is anything that needs to be specifically clarified.

**Comment by kitsonk** [comment]
Having an offline discussion on this had indicated that a bit of clarity should be provided.  So recommending that the style guide be updated to something like this:

> When using a *literal function*, an anonymous function that is being passed as an argument, arrow functions SHOULD be used.  Implicit returns from arrow functions are allowed if they increase the code readability and the return value is not ignored.


For example, acceptable:

```typescript
[ 1, 2, 3 ].forEach((value) => {
    console.log(value);
});
```

Unacceptable:

```typescript
[ 1, 2, 3 ].forEach((value) => console.log(value)); // return value is ignored
```

Acceptable:

```typescript
const arr = [ 1, 2, 3 ].map((value) => value * 2);
```
